src/interface/swarm.hpp

```diff
@@ -133,18 +178,50 @@ class Swarm {
   /// indicates gaps in the list.
   Real get_packing_efficiency() const { return num_active_ / (max_active_index_ + 1); }
 
-  bool mpiStatus;
-
   /// Remove particles marked for removal and update internal indexing
   void RemoveMarkedParticles();
 
   /// Open up memory for new empty particles, return a mask to these particles
-  ParArrayND<bool> AddEmptyParticles(const int num_to_add);
+  ParArrayND<bool> AddEmptyParticles(const int num_to_add, ParArrayND<int> &new_indices);
 
   /// Defragment the list by moving active particles so they are contiguous in
   /// memory
   void Defrag();
 
+  // used in case of swarm boundary communication
+  void SetupPersistentMPI();
+  std::shared_ptr<BoundarySwarm> vbswarm;
+  bool mpiStatus;
+  void allocateComms(std::weak_ptr<MeshBlock> wpmb);
+
+  int GetParticleDataSize() { return realVector_.size() + intVector_.size(); }
+
+  bool Send(BoundaryCommSubset phase);
+
+  bool Receive(BoundaryCommSubset phase);
+
+  vpack_types::SwarmVarList<Real> MakeRealList_(std::vector<std::string> &names);
+  vpack_types::SwarmVarList<int> MakeIntList_(std::vector<std::string> &names);
+
+  SwarmVariablePack<Real> PackVariablesReal(const std::vector<std::string> &names,
+                                            PackIndexMap &vmap);
+  SwarmVariablePack<Real> PackAllVariablesReal(PackIndexMap &vmap);
+  SwarmVariablePack<int> PackVariablesInt(const std::vector<std::string> &names,
+                                          PackIndexMap &vmap);
+
+  void PackAllVariables(SwarmVariablePack<Real> &vreal, SwarmVariablePack<int> &vint);
+  void PackAllVariables(SwarmVariablePack<Real> &vreal, SwarmVariablePack<int> &vint,
+                        PackIndexMap &rmap, PackIndexMap &imap);
```

Comment:
I saw where these were implemented above. I wonder if there's some way to reduce duplciated code between swarms and variable packs in general?

Also are you using these yet? If there were calls to them, I missed them.

Reply:
Yeah there's duplicated code for sure... but I don't immediately see what the best way to deal with that is. One example: `VariablePack`s are 3D in `i,j,k`, whereas `SwarmVariablePack`s are 1D in particle index. 

Yeah I use `PackAllVariables` in `Defrag`, for example, because I want to re-order every `ParticleVariable` in the same way. It cut the kernel calls in that part of the code way down.